compiler/src/iree/compiler/Dialect/Encoding/IR/EncodingAttrs.cpp

```diff
@@ -111,6 +116,63 @@ EncodingAttr EncodingAttr::clone(AffineMap bcastMap) {
              AffineMapAttr::get(bcastMap), getRoundDimsTo(), getLayouts());
 }
 
+Value EncodingAttr::calculateStorageSizeInBytes(Location loc,
+                                                OpBuilder &builder,
+                                                RankedTensorType type,
+                                                ValueRange dynamicDims) const {
+  SmallVector<OpFoldResult> shape =
+      getMixedValues(type.getShape(), dynamicDims, builder);
+  ArrayRef<int64_t> roundDimsTo = getRoundDimsToArray();
+  FailureOr<linalg::ContractionDimensions> cDims =
+      getEncodingContractionDims(*this);
+  auto pad = [&](int dim, int value) {
+    std::optional<unsigned> maybeMappedDim = mapDimToOperandIndex(dim);
+    if (!maybeMappedDim) {
+      return;
+    }
+    unsigned mappedDim = maybeMappedDim.value();
+    AffineExpr expr = builder.getAffineDimExpr(0);
+    shape[mappedDim] = affine::makeComposedFoldedAffineApply(
+        builder, loc, expr.ceilDiv(value) * value, {shape[mappedDim]});
+  };
+  for (auto m : cDims->m) {
+    pad(m, roundDimsTo[0]);
+  }
+  for (auto n : cDims->n) {
+    pad(n, roundDimsTo[1]);
+  }
+  for (auto k : cDims->k) {
+    pad(k, roundDimsTo[2]);
+  }
+
+  constexpr int64_t kNumBitsInByte = 8;
+  unsigned elementBits = IREE::Util::getTypeBitWidth(type.getElementType());
+  int64_t numBytesPerElem = 1;
+  if (elementBits > kNumBitsInByte) {
+    numBytesPerElem *=
+        IREE::Util::getRoundedElementByteWidth(type.getElementType());
```

Comment:
I'm don't want this dependency (encoding -> util) - util only has those functions because an encoding is not specified - this dialect should precisely specify the bit widths without needing to fall back to the auto detection behavior.

That is, it'd make more sense for the util functions to check if the type has an encoding and ask the encoding for the size instead of the encoding asking util to guess the size for something it should know. This may mean a bit of duplication of the util logic into here but it's simple and will be something we want to override for various encodings (block based encodings won't use that logic, for example, and we never want to have this code path calling into the generic util guessing code).

Reply:
Thanks for taking a look! I do have this dependency question. :)
I saw both solution, but I did not know which one is better. The logic should be simple, so duplication looks okay to me. Good to know that there is no such concern.